src/query_sink.rs

```diff
@@ -0,0 +1,163 @@
+use winapi::{
+    um::wbemcli::{
+        {IWbemClassObject,IWbemObjectSink, IWbemObjectSinkVtbl},
+        WBEM_NO_ERROR,
+    },
+    shared::{
+        ntdef::HRESULT,
+        wtypes::BSTR,
+        winerror::E_POINTER,
+    },
+    ctypes::{
+        c_long,
+    },
+};
+use com_impl::{ComImpl, VTable, Refcount};
+use log::{trace, warn};
+use std::ptr::NonNull;
+use wio::com::ComPtr;
+use async_channel::Sender;
+use crate::result_enumerator::IWbemClassWrapper;
+use crate::WMIError;
+
+/// Implementation for IWbemObjectSink.
+/// This [Sink] receives asynchronously the result of the query,
+/// through Indicate calls. When finished,the SetStatus method
+/// is called.
+/// [Sink]: https://en.wikipedia.org/wiki/Sink_(computing)
+/// # https://docs.microsoft.com/fr-fr/windows/win32/wmisdk/example--getting-wmi-data-from-the-local-computer-asynchronously
+#[repr(C)]
+#[derive(ComImpl)]
+#[interfaces(IWbemObjectSink)]
+pub struct QuerySink {
+    vtbl: VTable<IWbemObjectSinkVtbl>,
+    refcount: Refcount,
+    sender: Sender<Result<IWbemClassWrapper, WMIError>>,
+}
+
+impl QuerySink {
+    pub fn new(sender: Sender<Result<IWbemClassWrapper, WMIError>>) -> ComPtr<IWbemObjectSink> {
+        let ptr = QuerySink::create_raw(sender);
+        let ptr = ptr as *mut IWbemObjectSink;
+        unsafe { ComPtr::from_raw(ptr) }
+    }
+}
+
+#[com_impl::com_impl]
+unsafe impl IWbemObjectSink for QuerySink {
+    pub unsafe fn indicate(
+        &self,
+        lObjectCount: c_long,
+        apObjArray: *mut *mut IWbemClassObject
+    ) -> HRESULT {
+        trace!("Indicate call with {} objects", lObjectCount);
+        // TODO: Document when ObjectCount is <=0
+        if lObjectCount <= 0 {
+            return WBEM_NO_ERROR as i32;
+        }
+
+        let lObjectCount = lObjectCount as usize;
+        let tx = self.sender.clone();
+
+        unsafe {
+            // The array memory of apObjArray is read-only, and is owned by the caller of the Indicate method.
+            // The call to AddRef on each element of apObjArray to borrow is done by the IWbemClassWrapper::clone
+            // https://docs.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemobjectsink-indicate
+            for i in 0..lObjectCount {
+                // iterate over apObjArray elements
+                let p_el = *apObjArray.offset(i as isize);
```

Comment:
You can do `NonNull::new(*apObjArray.offset(i as isize))` and than match on the returned value